include/aws/io/dns.h

```diff
@@ -0,0 +1,602 @@
+ *   (3) root - Use the default providers (to resolve iterative queries).
+ *
+ * Returning to our original use cases:
+ *
+ *  (1) EzMode - Use an empty, unmodified provider config and get standard system default behavior (as if
+ *  getaddrinfo() was used) in all cases.
+ *
+ *  (2) Testing - Set bootstrap and default to a loopback address that has a mock listener attached to it.  Roots
+ *  can be set too to test iterative resolves.
+ *
+ *  (3) Privacy - Leave the bootstrap alone (use standard behavior to set up your default provider) but set the
+ *  default to something you trust (for example, a remote host that supports dns-over-https or dns-over-tls)
+ *
+ *  (4) Cache bypass - Add one or more root name servers to the root category.  They will be used as the start points
+ *  for resolving iterative queries (when the label cache does not have better starting information).
+ *
+ * Not yet solved (unsure if necessary):
+ *
+ *  (1) Do we need additional information on the provider record to support the
+ *  (unlikely) case of getting a truncated result and being unable to get around it via edns(0) options?  In
+ *  particular, would a fallback port for a tcp connection be useful and sufficient?
+ *
+ *  (2) Do we need to publicly expose the logic that works with resolv.conf/host.conf/nsswitch.conf?
+ */
+
+enum aws_dns_protocol {
+    AWS_DNS_PROTOCOL_UDP,
+    AWS_DNS_PROTOCOL_TCP,
+    AWS_DNS_PROTOCOL_TLS,
+    AWS_DNS_PROTOCOL_HTTPS,
+};
+
+struct aws_dns_service_provider_record {
+    enum aws_dns_protocol protocol;
+    struct aws_byte_cursor host;
+    uint16_t port;
+};
+
+struct aws_dns_service_provider_config {
+
+    /* If empty, the resolver will use resolv.conf-equivalent logic to select bootstrap providers */
+    struct aws_dns_service_provider_record *bootstrap_providers;
+    uint32_t bootstrap_provider_count;
+
+    /* If empty, the resolver will use the bootstrap providers for all recursive queries */
+    struct aws_dns_service_provider_record *default_providers;
+    uint32_t default_provider_count;
+
+    /* If empty, the resolver will use the default providers for iterative queries */
+    struct aws_dns_service_provider_record *root_providers; /* beginning of an array */
+    uint32_t root_provider_count;
+
+    /* If true, will skip all local host resolution rules (host.conf, /etc/hosts) */
+    bool disable_local_host_resolution;
+};
+
+/*
+ * Part 2 - aws_dns_resolver
+ * A thin interface around raw dns query/response.  A default resolver is included that uses
+ * aws_dns_service_provider_config to control what remote dns services are used.
+ *
+ * Musts/Requirements:
+ *   (1) 1 question allowed per query (aggregation is caller's responsibility)
+ *   (2) Implements retries (users should not layer retry on top)
+ *
+ * Shoulds:
+ *   (1) Cache authority/zones-of-responsibility in a label-based tree (iterative only?)
+ *   (2) Assume query input is a host name and not an IP address.
```

Comment:
would this break existing systems that rely on querying an ip address just returning that ip address?